### DISCUSSION TITLE
Moving some of the purge logic directly into the delete statement or chunking the work in smaller pieces to increase performance.

### DISCUSSION THREAD

**Discussion by EraYaN** [home-assistant/architecture#752] [General]
I currently have an install with about 80 million states and events (since about 2020-04) and purge has been failing (timeouts, tradfri killing home assistant, restarts etc) for a long time (hence the huge number of states and events). I asked on discord if an effort like this would be something that might make it into home assistant.

I was thinking that it would only really need to touch this file: https://github.com/home-assistant/core/blob/dev/homeassistant/components/recorder/purge.py 

So I was wondering if it would be possible to for example move some the the `SELECT` -> `DELETE` sql queries into a single `DELETE` query since at least for the states and events it seems to only really use the date as a condition.

Or alternatively to chunk the date window by day or something.
```
select min(last_updated) as oldest_date, 'states' as kind from states
union
select min(time_fired) as oldest_date, 'events' as kind from events;
```
for example is extremely fast because indexes exist, and that could give the code all it needs to run a bunch of select -> delete steps that should complete much much quicker. Since it would be able to use the last_updated index to quickly narrow down the actual rows.

Anyone any thoughts on how to do purges better/more resiliently?

Initial thoughts were (pseudo):
```python
oldest_date = sql min(last_update)
timechunk = 1h
target_date = purge_before
current_date = oldest_date
while current_date < target_date then
  sql start transaction
  ids = sql select ids with last_update between current_date - timechunk and current_date
  sql delete ids
  sql commit
  if count(ids) == MAX_ROWS_TO_PURGE: # Didn't get all of them, redo same timeframe
    continue
  current_date+=timechunk
  if count(ids)< MAX_ROWS_TO_PURGE/2: # Speedup if state over time density is not high
    timechunk *= 2 # or something much more clever that also adjusts down or something.
  
```

**Comment by bdraco** [comment]
How would that handle this case?
https://github.com/home-assistant/core/blob/dev/homeassistant/components/recorder/purge.py#L139

**Reply by EraYaN** [reply]
I mean if it would do a select -> delete set it would have those IDs, but really the database should handle this, and I think most RDBMSs will do so, and MSSQL well....

**Reply by EraYaN** [reply]
I might give it a try to see if it's at least faster to do it all directly. Maybe it's not even a worthy discussion, and it should just do select -> update ->delete chunks.

**Comment by bdraco** [comment]
The current process is purge a bit, wait a short bit, purge some more until done. 

The current design is intended to only purge small chunks at a time so it doesn't swamp the I/O on the system.  

It is important that everything remains responsive while a purge is happening.

**Comment by EraYaN** [comment]
So after adding a bunch of indexes, next thing I did is to add primary keys to the events and states tables. This plus the indexes help tremendously. (mostly on the time_fired and the last_updated colomn)
```sql
-- These both take about 8 minutes. (About 70-80 million rows each)
ALTER TABLE public.states ADD CONSTRAINT states_pk PRIMARY KEY (state_id);
ALTER TABLE public.events ADD CONSTRAINT events_pk PRIMARY KEY (event_id);

-- Repair all the broken foreign keys (takes about 30 minutes)
update states s set s.attributes_id = null where NOT EXISTS(select * from state_attributes sa where sa.attributes_id=s.attributes_id);
update states s set s.old_state_id = null where NOT EXISTS(select * from states s2 where s2.state_id=s.old_state_id);
update states s set s.event_id = null where NOT EXISTS(select * from events e where e.event_id=s.event_id);

-- These take anywhere from 8 to 35 minutes
ALTER TABLE public.states ADD CONSTRAINT states_fk FOREIGN KEY (attributes_id) REFERENCES state_attributes(attributes_id) ON DELETE CASCADE;
ALTER TABLE public.states ADD CONSTRAINT states_fk_old_state FOREIGN KEY (old_state_id) REFERENCES public.states(state_id) ON DELETE SET NULL;
ALTER TABLE public.states ADD CONSTRAINT states_fk_events FOREIGN KEY (event_id) REFERENCES public.events(event_id) ON DELETE CASCADE;
```

But I can't really benchmark the current code since after running for 48 hours+ it was just "too long" and I stopped it. I've now got it down to about 6 hours every 5 seconds. (About 51.5k events/states). Essentially the LIMIT doesn't stop processing the query quickly enough but if you make that purge before on a per 6 hour basis (or per day), is seems to work, although I'm letting it run through psql as pure SQL for now, and the above constraints handle all the updating of child rows.

But instead of running for days it now seems to complete in about 5 hours (including adding all the indexes). (From 2020-04-01 till today with about 200K-250k events/states per day). Good thing I'm doing this now since I'm about to add a bunch of chatty sensors and purging has been broken for a while now and hass still can't manage a single chunk in under 24 hours and can't stay alive for long enough either.

Next step is to test this on a much smaller database since this just all takes WAY to long. 

Also on the item of I/O, seems like even on hard drives (ZFS mirrors) postgres and chunked deletes are not able to really slow the server to a crawl. Although I'll do my smaller tests on SSDs for sure.

**Reply by bdraco** [reply]
It seems like these queries are not well optimized for postgresql or there expected indicies are missing in your setup or somehow didn't get created.  

The times you are quoting for deleting states and events are pretty wild as I can delete 10GiB of data from MariaDB in less than 30 minutes with the current code.

The recorder is definitely not tested with the same level of diligence using PostgreSQL as we do with MariaDB/MySQL and SQLite since there is no out of box set up offered for PostgreSQL.

**Reply by bdraco** [reply]
`ALTER TABLE public.states ADD CONSTRAINT states_fk FOREIGN KEY (attributes_id) REFERENCES state_attributes(attributes_id) ON DELETE CASCADE;`

Thats a many to one relationship to the `state_attributes` table.  Hopefully postgresql is smart enough to figure that out and not delete `attributes_id` from the `state_attributes` table if there is still at least one row in the `states` table that references it.

**Reply by EraYaN** [reply]
It would work the other way around IIRC with the ON DELETE. So if you remove the state_attributes item (the parent) then all children in the states table will be deleted. (Since otherwise it would be a broken foreign key constraint)

**Reply by EraYaN** [reply]
Also I think it might have been IO, the ZFS pool would only do about 2.5K-5K IOPS and for example when the autovacuum worker started halfway through it got real bad (10x slow down). So I disabled autovacuum during the removals and that helped a ton, add a VACUUM FULL for every say 30 GB and it actually was relatively fast. (about 5 hours). The smaller the database got the faster the removals. (Since at some point the database started fitting in RAM completely and the server only has 64GB which is shared with all other services running.)

**Reply by EraYaN** [reply]
Also when you say "out of the box setup" is that something I could contribute?

**Reply by bdraco** [reply]
If there are any slow queries or areas that poorly perform with PostgreSQL vs SQLite or MariaDB, we would be happy to accept PRs to `core` to improve the situation.

I'm not sure how much interest there is in an official PostgreSQL addon since we already have a MariaDB one (https://github.com/home-assistant/addons/blob/master/mariadb/README.md).  While it would make testing easier, I don't think there is much appetite for the additional maintainer-ship burden since we already offer two options.

**Reply by bdraco** [reply]
All the queries to find the old rows are now here 
https://github.com/home-assistant/core/blob/39ce25f76dd0ded7db2f1609fc457fe8a37fe8c1/homeassistant/components/recorder/queries.py#L581
